File: src/onedimensiondp/lc_746_min_climbing_stairs/min_climbing_stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: list[int]) -> int:
        minc = [0] * len(cost)
        minc[-1] = cost[-1]
        minc[-2] = cost[-2]

        for i in range(len(cost) - 3, -1, -1):
            minc[i] = min(minc[i+1], minc[i+2]) + cost[i]
        
        return min(minc[0], minc[1])

File: src/onedimensiondp/lc_746_min_climbing_stairs/min_climbing_stairs.java
import java.util.*;
class Solution {
    Map<Integer, Integer> map = new HashMap<>();

    public int minCostClimbingStairs(int[] cost) {
        return Math.min(dp(cost, 0), dp(cost, 1));
    }

    public int dp(int[] cost, int i) {
        if (map.containsKey(i))
            return map.get(i);
        if(i >= cost.length)
            return 0;
        int res = cost[i];
        res += Math.min(dp(cost, i+1), dp(cost, i+2));
        map.put(i, res);
        return res;
    }
}